big_tests/tests/sm_SUITE.erl

```diff
@@ -793,6 +803,117 @@ wait_for_resumption(Config) ->
+    escalus_connection:get_stanza(Bob, presence),
+
+    AliceSpec0 = escalus_fresh:create_fresh_user(Config, alice),
+    Resource = proplists:get_value(username, AliceSpec0),
+    AliceSpec = [{resource, Resource} | AliceSpec0],
+    start_hook_listener(Resource),
+    {ok, Alice, _} = escalus_connection:start(AliceSpec, ConnSteps ++ [stream_resumption]),
+    SMID = proplists:get_value(smid, Alice#client.props),
+    escalus_connection:send(Alice, escalus_stanza:presence(<<"available">>)),
+    escalus_connection:get_stanza(Alice, presence),
+    %% Ack the presence stanza
+    escalus:assert(is_sm_ack_request, escalus_connection:get_stanza(Alice, ack)),
+    escalus:send(Alice, escalus_stanza:sm_ack(1)),
+
+    escalus_connection:send(Bob, escalus_stanza:chat_to(common_helper:get_bjid(AliceSpec), <<"msg-1">>)),
+    escalus_connection:send(Bob, escalus_stanza:chat_to(common_helper:get_bjid(AliceSpec), <<"msg-2">>)),
+    %% kill alice connection
+    {ok, C2SPid} = get_session_pid(AliceSpec, Resource),
+    escalus_connection:kill(Alice),
+    wait_for_c2s_state_change(C2SPid, resume_session),
+    1 = length(get_user_alive_resources(AliceSpec)),
+
+    escalus:assert(is_chat_message, [<<"msg-1">>], verify_hook(0, Resource, 100)),
+    escalus:assert(is_chat_message, [<<"msg-2">>], verify_hook(0, Resource, 100)),
+    ?assertEqual(timeout, verify_hook(0, Resource, 100)),
+
+    %% send some messages and check if c2s can handle it
+    escalus_connection:send(Bob, escalus_stanza:chat_to(common_helper:get_bjid(AliceSpec), <<"msg-3">>)),
+    escalus_connection:send(Bob, escalus_stanza:chat_to(common_helper:get_bjid(AliceSpec), <<"msg-4">>)),
+    escalus:assert(is_chat_message, [<<"msg-3">>], verify_hook(0, Resource, 100)),
+    escalus:assert(is_chat_message, [<<"msg-4">>], verify_hook(0, Resource, 100)),
+    ?assertEqual(timeout, verify_hook(0, Resource, 100)),
+
+    %% alice comes back and receives unacked message
+    {NewResource, NewAlice} =
+        case TestCase of
+            unacknowledged_message_hook_resume ->
+                unacknowledged_message_hook_resume(AliceSpec, Resource, SMID);
+            unacknowledged_message_hook_offline ->
+                unacknowledged_message_hook_offline(AliceSpec, Resource, ConnSteps, C2SPid);
+            unacknowledged_message_hook_bounce ->
+                unacknowledged_message_hook_bounce(AliceSpec, Resource, ConnSteps, C2SPid)
+        end,
+
+    mongoose_helper:wait_until(
+        fun() ->
+            Stanza = escalus_connection:get_stanza(NewAlice, msg),
+            escalus:assert(is_chat_message, [<<"msg-4">>], Stanza),
+            ok
+        end, ok),
+
+    {ok, NewC2SPid} = get_session_pid(AliceSpec, NewResource),
+    escalus_connection:kill(NewAlice),
+    wait_for_c2s_state_change(NewC2SPid, resume_session),
+
+    mongoose_helper:wait_until(
+        fun() ->
+            escalus:assert(is_chat_message, [<<"msg-1">>], verify_hook(1, NewResource, 100)),
```

Comment:
`verify_hook` has side effect of removing a message from the queue, so if this fun doesn't succeed on the first try, then it will never succeed, so `wait_until` doesn't help in anything here.

Reply:
we verify the sequence here, and obviously `wait_until` added because of the reason. in some testcases we also have presence stanza at the beginning.

Reply:
however, that is good idea to filter out only messages for this hook.